Exercise 8

cc1000 opened the pull request.

cc1000 committed:
Push release image to Docker Hub (WIP)

cc1000 committed:
Add correct image/tag name to production build

cc1000 committed:
Move publish/deploy to script file and invoke from deploy step

cc1000 committed:
Log in to Docker Hub before starting Docker builds

cc1000 committed:
Amend Dockerfile to support running production image in Heroku

cc1000 committed:
Add Heroku deployment

cc1000 committed:
Rename Heroku API key environment variable; use stdin for Heroku Docker login

cc1000 committed:
Restrict deployment to master branch

JackMeadDev commented on build_scripts/publish_and_deploy.sh:
```diff
@@ -0,0 +1,5 @@
+docker push $1/todo-app:latest
```
Minor, but if we moved these first couple of lines into e.g. a before_deploy stage, then we'd be able to refer to the variables with names rather than just argument positions

JackMeadDev commented on Dockerfile:
```diff
@@ -13,8 +13,9 @@ COPY poetry.lock pyproject.toml /
 FROM base as production
-ENTRYPOINT poetry run gunicorn -b 0.0.0.0:5000 --chdir /src wsgi
-RUN poetry install --no-dev --no-root
+ENTRYPOINT ["poetry run gunicorn -b 0.0.0.0:$PORT --chdir /src wsgi"]
```
Two thoughts here:
- This change from a hardcoded port value means that your instructions for running the prod container locally aren't valid any longer & the container fails to start, we should fix that (probably either by defining a default port value here in the dockerfile, or by passing one in from `docker-compose.yml`)
- The exec form of entrypoint expects an array of strings passing the command and all arguments separately, and indeed is causing another layer of issues when running this. It will also fail to substitute the `$PORT` variable - one solution here would be to start a bash script from entrypoint that then runs the start command for the container

JackMeadDev reviewed (changes requested):
Nice work, I can see your docker image on docker hub, your heroku app deployed and the pipeline looks good. There's a couple of issues with running your updated dockerfile locally for the prod setup though which we should resolve before merging

cc1000 committed:
Move Docker publish to before_deploy stage

cc1000 committed:
Setting branch filter to feature branch for testing

cc1000 committed:
Pass in PORT environment variable from prod docker-compose; move production entrypoint to bash script

cc1000 commented:
@JackMead, I believe I've addressed your comments above. Going ahead with merge for the next project exercise but please let me know if there are any further issues to address.

JackMeadDev commented on build_scripts/publish_and_deploy.sh:
```diff
@@ -1,5 +1,3 @@
-docker push $1/todo-app:latest
-docker tag $1/todo-app:latest registry.heroku.com/cree1000-todo-app/web
-echo $3 | docker login -u $2 --password-stdin registry.heroku.com
+echo $2 | docker login -u $1 --password-stdin registry.heroku.com
```
I was thinking this line could also be moved to save accessing any variables by position in this file, but it's not particularly important

JackMeadDev reviewed (commented):
This looks good, and I was able to run your prod compose file locally again without issue. Happy that this is merged
